app/src/main/java/fr/free/nrw/commons/contributions/ContributionsListAdapter.java

```diff
@@ -63,7 +85,63 @@ public void bindView(View view, Context context, Cursor cursor) {
                 views.stateView.setVisibility(View.VISIBLE);
                 views.stateView.setText(R.string.contribution_state_failed);
                 views.progressView.setVisibility(View.GONE);
+                views.retryButton.setVisibility(View.VISIBLE);
+                views.cancelButton.setVisibility(View.VISIBLE);
+
+                views.retryButton.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View view) {
+                        retryUpload(cursor);
+                    }
+                });
+
+                views.cancelButton.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View view) {
+                        deleteUpload(cursor);
+                    }
+                });
+
+
                 break;
         }
     }
+
+    /**
+     * Retry upload when it is failed
+     * @param cursor cursor will be retried
+     */
+    public void retryUpload(Cursor cursor) {
+        if (NetworkUtils.isInternetConnectionEstablished(mContext)) {
+            Contribution c = contributionDao.fromCursor(cursor);
+            if (c.getState() == STATE_FAILED) {
+                uploadService.queue(UploadService.ACTION_UPLOAD_FILE, c);
+                Timber.d("Restarting for %s", c.toString());
+            } else {
+                Timber.d("Skipping re-upload for non-failed %s", c.toString());
+            }
+        } else {
+            ViewUtil.showLongToast(mContext,R.string.this_function_needs_network_connection);
+        }
+
+    }
+
+    /**
+     * Delete a failed upload attempt
+     * @param cursor cursor which will be deleted
+     */
+    public void deleteUpload(Cursor cursor) {
+        if (NetworkUtils.isInternetConnectionEstablished(mContext)) {
```

Comment:
This deletes the thing from the local db right, do we need to have a check for Network Connection ?

Reply:
Very valid concern but yes, since we can not make sure db is up to date. If we network off during upload, then click to retry or cancel, cursor displays latest uploaded item and it says "Skipping deletion for non-failed contrib [previous contrib]" . It points to wrong contrib. For retry it was same. You can go 
f3818c6  commit and see. Thanks for reviewing though:)

Reply:
Ohh, if that is the case, then yeah it is required